aten/src/ATen/native/BlasKernel.cpp

```diff
@@ -215,6 +215,56 @@ static inline float16_t reduce(float16x8_t x) {
         return reduce(vadd_f16(vget_low_f16(x), vget_high_f16(x)));
 }
 
+#define F16_ELEMENTS_PER_ITERATION 32
+#define F16_ELEMENTS_PER_REGISTER 8
+#define F16_REGISTERS_PER_ITERATION (F16_ELEMENTS_PER_ITERATION / F16_ELEMENTS_PER_REGISTER)
+static inline double reduce(float16x8_t x[F16_REGISTERS_PER_ITERATION]) {
+  int offset = F16_REGISTERS_PER_ITERATION / 2;
+  for (int i = 0; i < offset; ++i) {
+    x[i] = vaddq_f16(x[i], x[offset + i]);
+  }
+  offset /= 2;
+  for (int i = 0; i < offset; ++i) {
+    x[i] = vaddq_f16(x[i], x[offset + i]);
+  }
+  offset /= 2;
+  for (int i = 0; i < offset; ++i) {
+    x[i] = vaddq_f16(x[i], x[offset + i]);
+  }
+  const float32x4_t t0 = vcvt_f32_f16(vget_low_f16(x[0]));
+  const float32x4_t t1 = vcvt_f32_f16(vget_high_f16(x[0]));
+  return (double)vaddvq_f32(vaddq_f32(t0, t1));
+
+}
+
+#ifdef __ARM_FEATURE_FMA
+#define F16_FMA(a, b, c) vfmaq_f16((a), (b), (c))
+#else
+#define F16_FMA(a, b, c) vaddq_f16((a), vmulq_f16((b), (c)))
+#endif
```

Comment:
Nit (less preprocessor code and more inline functions is better imo)
Also, aren't fma(a, b, c) == a*b + c ? I.e. order of operands in your non-fma implementaiton is wrong
```suggestion
static inline f16_fma(float16x8_t a, float16x8_t b, float16x8_t c) {
#ifdef __ARM_FEATURE_FMA
 return vfmaq_f16(a, b, c);
#else
 return vaddq_f16(a, vmulq_f16(b, c));
#endif
}
```